sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -5855,51 +5865,79 @@ class AstBuilder extends DataTypeAstBuilder
     if (!SQLConf.get.getConf(SQLConf.OPERATOR_PIPE_SYNTAX_ENABLED)) {
       operationNotAllowed("Operator pipe SQL syntax using |>", ctx)
     }
-    // This helper function adds a table subquery boundary between the new operator to be added
-    // (such as a filter or sort) and the input plan if one does not already exist. This helps the
-    // analyzer behave as if we had added the corresponding SQL clause after a table subquery
-    // containing the input plan.
-    def withSubqueryAlias(): LogicalPlan = left match {
-      case s: SubqueryAlias =>
-        s
-      case u: UnresolvedRelation =>
-        u
-      case _ =>
-        SubqueryAlias(SubqueryAlias.generateSubqueryName(), left)
-    }
-    Option(ctx.selectClause).map { c =>
-      withSelectQuerySpecification(
-        ctx = ctx,
-        selectClause = c,
-        lateralView = new java.util.ArrayList[LateralViewContext](),
-        whereClause = null,
-        aggregationClause = null,
-        havingClause = null,
-        windowClause = null,
-        relation = left,
-        isPipeOperatorSelect = true)
-    }.getOrElse(Option(ctx.whereClause).map { c =>
-      withWhereClause(c, withSubqueryAlias())
-    }.getOrElse(Option(ctx.pivotClause()).map { c =>
-      if (ctx.unpivotClause() != null) {
-        throw QueryParsingErrors.unpivotWithPivotInFromClauseNotAllowedError(ctx)
+
+    // Helper function to extract window definitions from the plan
+    def extractWithWindowDefinitions(
+        plan: LogicalPlan): (LogicalPlan, List[WithWindowDefinition]) = {
+      var currentPlan = plan
+      var withWindowDefinitions = List.empty[WithWindowDefinition]
+      while (currentPlan.isInstanceOf[WithWindowDefinition]) {
+        val wwd = currentPlan.asInstanceOf[WithWindowDefinition]
+        withWindowDefinitions = withWindowDefinitions :+ wwd
+        currentPlan = wwd.child
       }
-      withPivot(c, left)
-    }.getOrElse(Option(ctx.unpivotClause()).map { c =>
-      if (ctx.pivotClause() != null) {
-        throw QueryParsingErrors.unpivotWithPivotInFromClauseNotAllowedError(ctx)
+      (currentPlan, withWindowDefinitions)
+    }
+
+    // Extract the base child and the list of WithWindowDefinition nodes
+    val (baseChild, withWindowDefinitions) = extractWithWindowDefinitions(left)
+
+    // Helper function to process the right side of the operator pipe
+    def processRightSide(relation: LogicalPlan): LogicalPlan = {
```

Comment:
same, this is only called once on L5936. We can just assign it to a `val` instad.